---
title: "Gene Annotation"
author: "Fatemeh"
date: "June 17, 2019"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## FinalGeneSetAnnotation.R
This script has the fallowing functions:\
\
\textbf{0. annotate.final.geneset.round1\(\)}\
\textbf{Input}: \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/GenePrediction/Integrate_Tse_Ara.R}{\textbf{integrated\_tse\_ara.txt}}\
It reads the main integrated gene file from integrated_tse_ara.txt\
Keeps genes with either their ara score is $>$ 106 or their tse score is $>$ 49\
Column \textbf{genefun} is added to the genefile table and filled in the fallowing order:
\begin{itemize}
\item[1.] genes with same tse and ara identity are assign the same idenitity\
\item[2.] genes with different identity, pseudo|truncated genes, genes with un assigned identity and genes with letter N in their sequence are shown as \#\
\end{itemize}

Column \textbf{note} is added and fill it in the fallowing order:
\begin{itemize}
\item[1.] genes with the same ara and tse marked as "T"\
\item[2.] genes with unassigned identity by both tse and ara are set as "UnAssigned"\
\item[3.] genes with letter N in their sequence are set as "ContainsN"\
\item[4.] genes with unmatched identity between ara and tse are set as "Undet" \
\end{itemize}
these top 4 cases had no overlap!\
I kept track of the remained number of genes at each step and wrote it as comments.\
\
```{r , echo=FALSE,include=FALSE}
library(gridExtra)
library(seqinr)
library(gdata)
library(reshape2)
library(plyr)
library(gsubfn)
library(ggplot2)
library(seqinr)
#source("https://bioconductor.org/biocLite.R")
library(Biostrings)

annotate.final.geneset.round1 <- function() {

  genefilepath <-
  genefilepath <-
    "/home/fatemeh/TriTrypGenome/GenePrediction/integrated_tse_ara.txt"
  genefile <-
    read.table(genefilepath, header = TRUE, colClasses = "character")
  #table(genefile$foundby)
  #ara both  tse 
  #750 3597   34 
  # set the ara score of genes not found by aragorn/tse to -1 to convert the column to integer
  genefile[genefile$arascore == "notfound",]$arascore <- -1
  genefile$arascore <- as.integer(genefile$arascore)
  genefile[genefile$tsescore == "notfound",]$tsescore <- -1
  genefile$tsescore <- as.integer(genefile$tsescore)
  # cuttoff score for TSE 50, for ARA 107
  dismiss0 <-
    (genefile$foundby == "both") &
    (genefile$arascore < 107 & genefile$tsescore < 50)
  dismiss1 <-
    (genefile$foundby == "ara") &
    (genefile$arascore < 107) # most of these genes were pseudo or truncated or both
  dismiss2 <- (genefile$foundby == "tse") & (genefile$tsescore < 50)
  dismissedGenes <- genefile[(dismiss0 | dismiss1 | dismiss2),]
  #table(dismissedGenes$foundby)
  #ara both  tse
  #714   18   33(28 of these tse only genes were pseudo| truncated genes)
  
  Anotated_genefile <-
    genefile[(!dismiss0 & !dismiss1 & !dismiss2),]
  #ara both  tse
  #36 3579    1
  
  # genefunc column is added which shows the presented gene identity in table
  # Genes marked as ?? in this column include: pseudo|truncated genes, genes with unmatched identity,
  # genes with unassigned identity|anticodon by any of genefinders, and genes with letter N|n in their sequence.
  
  ambiguty1 <-
    Anotated_genefile$tsefunc == "" &
    Anotated_genefile$arafunc == "" 
  ambiguty2 <-
    Anotated_genefile$tsenote != "notfound" # genes that are not markes as notfound by tse are pseudo|truncated
  ambiguty3 <-
    Anotated_genefile$foundby == "both" &
    Anotated_genefile$tsefunc != Anotated_genefile$arafunc 
  ambiguty4 <- logical(length = nrow(Anotated_genefile))
  for (i in 1:nrow(Anotated_genefile)) {
    if (Anotated_genefile$foundby[i] != "ara")
      ambiguty4[i] <-
        length(grep("n|N", Anotated_genefile$tsegeneseq[i])) == 1
    else
      ambiguty4[i] <-
        length(grep("n|N", Anotated_genefile$arageneseq[i])) == 1
  }
  
  ambiguties <- ambiguty1 | ambiguty2 | ambiguty3 | ambiguty4
  #table(Anotated_genefile[!ambiguties, ]$foundby)
  #ara both
  #36  3535
  Anotated_genefile$genefunc <- ""
  Anotated_genefile[!ambiguties,]$genefunc <-
    Anotated_genefile[!ambiguties,]$arafunc
  Anotated_genefile[ambiguties,]$genefunc <- "#"
  # 45 genes are marked as #from which 1 found by tse and 44 found by both
  # 3571 not marked as #
  # resolving the identity of some of the # genes
  # column note is added for the status of gene's identity
  Anotated_genefile$note <- "T"
  Anotated_genefile[ambiguty1, ]$note <- "UnAssigned" # 2 genes
  Anotated_genefile[ambiguty2, ]$note <-
    Anotated_genefile[ambiguty2, ]$tsenote # 4 truncated 2 pseudo
  Anotated_genefile[ambiguty4, ]$note <- "ContainsN" # 4 genes
  # top 3 lines do not overlap
  Anotated_genefile[ambiguty3, ]$note <-
    "Undet" # 33 genes found by both with unmatch identity
  #table(paste(Anotated_genefile[ambiguty3, ]$tsefunc,Anotated_genefile[ambiguty3, ]$arafunc,sep="|"))
  #E|L G|W I|D ?|L M|L M|O R|S Y|N 
  #1   3   3   3   9   2   1  11 
  
  # based on the results from annotate.final.geneset.round2() in complete version of this script (FinalGeneSetAnnotation) we will add 10 Y genes which have same folding by both genefinders
  # These genes have relatively higher tse score than ara score. tse score of 72 (based on density figure it is amongst well predicted genes). 
  # and ara score of 108 which is very close to the aragorns lowscore genes.
  # based on the article: Gene organization and sequence analyses of transfer RNA genes in Trypanosomatid parasites
  # Analysis of the tRNA genes in Tritryps indicated that only the tRNA-Tyr genes contain an intron; which was previously reported 
  # in T. brucei. The intron is 11 bases long in L. major and T. brucei, and 13 bases long in T. cruzi; and as in other organisms, 
  # it is located between bases 37 and 38 (data not shown).
  # based on the TSE annotation the introns of length 13 are annotated between position 37 and 38, however it is annotated at site 35-36
  Ygenes <- Anotated_genefile$genefunc == "#" &
    Anotated_genefile$tsefunc == "Y" &
    Anotated_genefile$arascore > 106 &
    Anotated_genefile$tsescore > 50
  Anotated_genefile[Ygenes, ]$genefunc <- Anotated_genefile[Ygenes, ]$tsefunc
  Anotated_genefile[Ygenes, ]$note <- "resolved"
  Anotated_genefile
}
Anotated_genefile <- annotate.final.geneset.round1()
```
\textbf{1. genome.nuc.composition\(\):}\
\textbf{Input}: \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/GenomeNucComposition.txt}{\textbf{GenomeNucComposition.txt}} and \textbf{output of function annotate.final.geneset.round1\(\)}\
\textbf{output}: output is saved in latex format in this file \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/Document_Latex/GenomeNucComposition_Latex.txt}{\textbf{GenomeNucComposition\_Latex.txt}}\
\
The complete script for calculating the nucleotide composition is in FinalGeneSetAnnotation.R. Reading 46 genomes and caclulating the compositions is a bit time consuming, so I have already done that part in function genome.nuc.composition() in the main R script and saved the table as GenomeNucComposition.txt. Here, in function genome.nuc.composition2() I only read file GenomeNucComposition.txt, and marge it with a new column called genecounts which I calculate from the output of function annotate.final.geneset.round1(). In the end the file is written in latex format.\
```{r , echo=FALSE}
genome.nuc.composition2 <- function(){
  genefile <- annotate.final.geneset.round1()
  genecountdf <- as.data.frame(table(genefile$sourceOrg))
  genecountdf$Var1 <- as.character(genecountdf$Var1)
  genecountdf$Freq <- as.integer(genecountdf$Freq)
  #genomefile <- genome.nuc.composition
  filepath <- "/home/fatemeh/TriTrypGenome/GenomeNucComposition.txt"
  genomefile <-
    read.table(filepath, header = TRUE, colClasses = "character")
  genomefile$seqcounts <- as.integer(genomefile$seqcounts)
  genomefile  <- genomefile[order(genomefile$seqcounts), ]
  genomefile$genecounts <- 0
  for (i in 1:nrow(genomefile)) {
    currname <- genomefile$organism_shortname[i]
    if (currname == "LmajorSD75.1")
      currname <- "LmajorSD75"
    genomefile$genecounts[i] <-
      genecountdf[genecountdf$Var1 == currname, ]$Freq
  }
  
  # remove the column ATp and organism_longname from the table
  genomefile2 <-
    genomefile[, !(names(genomefile) %in% c("ATp", "organism_longname"))]
  genomefile2  <- genomefile2[order(genomefile2$genecounts), ]
  
  # add the column genesMissing ????
  # write latex file
  firstline <-
    c("organism",
      "Aperc",
      "Tperc",
      "Cperc",
      "Gperc",
      "GCp",
      "seqcounts",
      "genecounts")
  Llines <- paste(firstline, collapse = "&")
  Llines <- paste(Llines, "\\\\", sep = "")
  for (i in 1:nrow(genomefile2)) {
    currline <-
      paste(genomefile2[, 1:ncol(genomefile2)][i, ], collapse = "&")
    currline <- paste(currline, "\\\\", sep = "")
    Llines <- c(Llines, currline)
  }
  genomefile2
  # writeLines(Llines,
  #            "/home/fatemeh/TriTrypGenome/GenomeNucComposition_Latex.txt",
  #            sep = "\n")
}
nuc_comp_table <- genome.nuc.composition2()

print(nuc_comp_table[!(names(nuc_comp_table) %in% c("seqcounts"))],row.names = FALSE)
```
\textbf{2. Score.visualization\(\)}\
\textbf{Input}: \textbf{Input}: \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/GenePrediction/Integrate_Tse_Ara.R}{\textbf{integrated\_tse\_ara.txt}}\
\
This functions read the main integrated gene file integrated_tse_ara.txt and shows the distribution of gene scores \
```{r plot, echo=FALSE}

geneDF <- read.table("/home/fatemeh/TriTrypGenome/GenePrediction/integrated_tse_ara.txt",header = TRUE)
geneDF2 <- geneDF[geneDF$foundby == "ara" | geneDF$foundby == "both",]
geneDF2$arascore <- as.double(geneDF2$arascore)
geneDF2$Foundby <- "Aragorn-only"
geneDF2[geneDF2$foundby!="ara",]$Foundby <- "Both"
geneDF2[geneDF2$tsenote!="notfound",]$Foundby <- "Pseudo/Truncated"
geneDF2$Foundby <- factor(geneDF2$Foundby,levels=c("Both","Aragorn-only","Pseudo/Truncated"))
#"Aragorn score of genes found by both genefinders TSE and ARA, and genes found by only ARA"
araplot <- ggplot(geneDF2,
         aes(
           x = arascore,
           fill = Foundby
         )) + theme_bw() +
    geom_density(alpha=0.3) + xlab("Aragorn Score") + ggtitle("Aragorn Score Distributions") +    theme(legend.title=element_blank())

geneDF3 <- geneDF[geneDF$foundby == "tse" | geneDF$foundby == "both",]
geneDF3$tsescore <- as.double(geneDF3$tsescore)
geneDF3$Foundby <- "tRNAscan-SE2-only"
geneDF3[geneDF3$foundby =="both",]$Foundby <- "Both Gene-Finders"
geneDF2$Foundby <- factor(geneDF2$Foundby,levels=c("Both Gene-Finders","tRNAscan-SE2-only"))
tseplot <- ggplot(geneDF3,
         aes(
           x = tsescore,
           fill = Foundby
         )) + theme_bw() +
    geom_density(alpha=0.3) + xlab("tRNAscan-SE2 Score") + ggtitle("tRNAscan-SE2 Score Distributions") + theme(legend.title=element_blank())
grid.arrange(araplot, tseplot, ncol=1)
```
\newpage
```{r summary, echo=FALSE}
gcContent <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("G", "C")]) * 100
}
atContent <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("A", "T")]) * 100
}
Apercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("A")]) * 100
}
Cpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("C")]) * 100
}
Gpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("G")]) * 100
}
Tpercentage <- function(x) {
  x <- DNAString(toupper(x))
  alf <- alphabetFrequency(x, as.prob = TRUE)
  sum(alf[c("T")]) * 100
}
```
\textbf{3. create.summary.table\(\)}\
\textbf{Input}: output of function annotate.final.geneset.round1()\
\textbf{output}: output is saved in latex format in this file \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/Document_Latex/SummaryTable_Latex.txt}{\textbf{SummaryTable\_Latex.txt}}\
\
Read the annotated gene file (output of function annotate.final.geneset.round1) and Creates a summary table of genes after the score filtering.\
Last three columns are each for one gene set. Sets are defined as: a) Intersection Of two gene finders. Two genes are considered same gene if their coordinate overlaps at least one base. Displacement of overlapped genes between ARA and TSE does not pass 4bp. b) Union of two gene finders. c) Genes found by only ARA. Genes marked as # include: pseudo|truncated genes (6 genes), genes with different predicted identity by two genefinders(23 genes), genes with unassigned identity|anticodon by any of genefinders (2 genes), and genes with letter N in their sequence(we had 4 of these genes). we also had 1 genes preicted by only TSE labeled as # which is not shown in this table as a seperate column, however it is considered in the union set.\
```{r , echo=FALSE}
#create.summary.table <- function(genefile) {
  genefile <- annotate.final.geneset.round1()
  # create a table with columns: Annotation, Intersection, ARAonly, Union
  firstcol <- c("#tRNA","#N/#G","Min Gene Length","Max Gene Length","%intron","%G" ,"%C" ,"%T" ,"%A","A",  "C",  "D",  "E",  "F",  "G",  "H",  "I",  "K",  "L",  "M",  "N",  "P",  "Q",  "R",  "S",  "T",  "V",  "W",  "X",  "Y",  "Z", "#")
  resulttable <- data.frame(firstcol,rep(0,length(firstcol)),rep(0,length(firstcol)),rep(0,length(firstcol)))
  names(resulttable) <- c("Annotation", "Intersection", "ARAonly", "Union")
  
  isboth <- genefile$foundby=="both"
  bothdf <- genefile[isboth,]
  isaraonly <- genefile$foundby=="ara"
  intersectDF <- genefile[isboth,]
  ARAonlyDF <- genefile[isaraonly,]
  genefile$geneseq <- ""
  genefile[isaraonly,]$geneseq <- genefile[isaraonly,]$arageneseq
  genefile[!isaraonly,]$geneseq <- genefile[!isaraonly,]$tsegeneseq
  
  resulttable[1,2:4] <- c(nrow(intersectDF),nrow(ARAonlyDF),nrow(genefile))
  resulttable[2,2:4] <- c(sum(nchar(intersectDF$tsegeneseq))/nrow(intersectDF),sum(nchar(ARAonlyDF$arageneseq))/nrow(ARAonlyDF),sum(nchar(genefile$geneseq))/nrow(genefile)) # sum of gene length / # genes from the first row
  resulttable[3,2:4] <- c(min(nchar(intersectDF$tsegeneseq)),min(nchar(ARAonlyDF$arageneseq)),min(nchar(genefile$geneseq)))
  resulttable[4,2:4] <- c(max(nchar(intersectDF$tsegeneseq)),max(nchar(ARAonlyDF$arageneseq)),max(nchar(genefile$geneseq)))
  resulttable[5, 2:4] <-
    c(
      100 * nrow(bothdf[bothdf$tseintronbegin != 0, ]) / nrow(bothdf),
      100 * nrow(ARAonlyDF[ARAonlyDF$araintronbegin != "nointron", ]) / nrow(ARAonlyDF),
      100 * (nrow(bothdf[bothdf$tseintronbegin != 0, ]) + nrow(ARAonlyDF[ARAonlyDF$araintronbegin !=
                                                                           "nointron", ])) / nrow(genefile)
    )
  resulttable[6,2:4] <- c(Gpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Gpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Gpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[7,2:4] <- c(Cpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Cpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Cpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[8,2:4] <- c(Tpercentage(paste(intersectDF$tsegeneseq,collapse = "")),Tpercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Tpercentage(paste(genefile$geneseq,collapse = "")))
  resulttable[9,2:4] <- c(Apercentage(paste(intersectDF$tsegeneseq,collapse = "")),Apercentage(paste(ARAonlyDF$arageneseq,collapse = "")),Apercentage(paste(genefile$geneseq,collapse = "")))
  
  # make a table of Class frequencies for each set:
  intersect_ClassFreq <- as.data.frame(table(intersectDF$genefunc))
  names(intersect_ClassFreq) <- c("class","freq")
  AraOnly_ClassFreq <- as.data.frame(table(ARAonlyDF$genefunc))
  names(AraOnly_ClassFreq) <- c("class","freq")
  Union_ClassFreq <- as.data.frame(table(genefile$genefunc))
  names(Union_ClassFreq) <- c("class","freq")
  intersect_ClassFreq$class <- as.character(intersect_ClassFreq$class)
  AraOnly_ClassFreq$class <- as.character(AraOnly_ClassFreq$class)
  Union_ClassFreq$class <- as.character(Union_ClassFreq$class)
  
  for (i in 1:nrow(intersect_ClassFreq)) {
    curr_class <- intersect_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,2] <- intersect_ClassFreq$freq[i]
  }
  for (i in 1:nrow(AraOnly_ClassFreq)) {
    curr_class <- AraOnly_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,3] <- AraOnly_ClassFreq$freq[i]
  }
  for (i in 1:nrow(Union_ClassFreq)) {
    curr_class <- Union_ClassFreq$class[i]
    resulttable[resulttable$Annotation==curr_class,][,4] <- Union_ClassFreq$freq[i]
  }
  
  resulttable$Intersection <- round(resulttable$Intersection,digits = 0)
  resulttable$ARAonly <- round(resulttable$ARAonly,digits = 0)
  resulttable$Union <- round(resulttable$Union,digits = 0)
  resulttable$Annotation <- as.character(resulttable$Annotation)
  #Latex.file.prep(resulttable,"resulttable")
  
  print(resulttable,row.names = FALSE)

#}
```
\newpage
```{r ,echo=FALSE, include=FALSE}
create.clusterDF <- function(tsfmGeneDF){
  geneset <- assignCluster(tsfmGeneDF,1000)
  m <- matrix(nrow = max(geneset$cluster),ncol = 9)
  clusterDF<- as.data.frame(m)
  names(clusterDF) <- c("sourceOrg","soureSeq","clusterID","clusterSeq1","clusterSeq2", "clusterBegin","clusterEnd","clusterDir","clusterLen")
  for (i in 1:max(geneset$cluster)) {
    clusterDF$sourceOrg[i] <- geneset[geneset$cluster==i,]$sourceOrg[1]
    clusterDF$soureSeq[i] <- geneset[geneset$cluster==i,]$sourceseq[1]
    clusterDF$clusterID[i] <- i
    clusterDF$clusterSeq1[i] <- paste(geneset[geneset$cluster==i,]$genecode,collapse = ",")
    clusterDF$clusterSeq2[i] <- paste(geneset[geneset$cluster==i,]$genefunc,collapse = "")
    clusterDF$clusterBegin[i] <- sort(geneset[geneset$cluster==i,]$begin)[1]
    clusterDF$clusterEnd[i] <- sort(geneset[geneset$cluster==i,]$end)[nrow(geneset[geneset$cluster==i,])]
    clusterDF$clusterDir[i] <- paste(geneset[geneset$cluster==i,]$direction,collapse = "")
    clusterDF$clusterLen[i] <- nrow(geneset[geneset$cluster==i,])
  }
  clusterDF$class <- toupper(substr(clusterDF$sourceOrg,1,1))
  clusterDF[clusterDF$class!="T" & clusterDF$class!="L",]$class <- "O"
  clusterDF
}
assignCluster <- function(geneDF, clusterdistance) {
  
  geneDF$tsebegin <- as.integer(geneDF$tsebegin)
  geneDF$tseend <- as.integer(geneDF$tseend)
  geneDF$arabegin <- as.integer(geneDF$arabegin)
  geneDF$araend <- as.integer(geneDF$araend)
  geneDF$begin <- geneDF$tsebegin
  geneDF[geneDF$foundby=="ara",]$begin <- geneDF[geneDF$foundby=="ara",]$arabegin
  geneDF$end <- geneDF$tseend
  geneDF[geneDF$foundby=="ara",]$end <- geneDF[geneDF$foundby=="ara",]$araend
  
  geneDF <-
    geneDF[order(geneDF$sourceOrg, geneDF$sourceseq, geneDF$begin), ]
  geneDF$cluster <- 0
  geneDF$cluster[1] <- 1
  setnumber <- 1
  
  for (i in 1:(nrow(geneDF) - 1)) {
    geneDist <-
      abs(geneDF$begin[i + 1] - geneDF$begin[i])
    if ((geneDist < clusterdistance) &&
        (geneDF$sourceOrg[i] == geneDF$sourceOrg[i + 1]) &&
        (geneDF$sourceseq[i + 1] == geneDF$sourceseq[i]))
      geneDF$cluster[i + 1] <- setnumber
    else
    {
      setnumber <- setnumber + 1
      geneDF$cluster[i + 1] <- setnumber
    }
  }
  geneDF
}
```
\textbf{4. clustersize.dist.visualize\(\)}\
\textbf{Input}: output of function annotate.final.geneset.round1()\
Read the annotated gene file (output of function annotate.final.geneset.round1) and visualizes the Cluster size distribution for three categories of TryTryp genomes. Labels in green on top of each bar show the percentage of total number of genes as cluster of a specific length. Each color refers to one category of TriTryp genomes. Numbers within each color section of the bar shows the counts of clusters with a specific length.\
```{r ,echo=FALSE, warning=FALSE}
#clustersize.dist.visualize <- function(genefile){
  genefile <- annotate.final.geneset.round1()
  clusterdf <- create.clusterDF(genefile)
  # dataframe c("clusterLen","Tfreq","Lfreq","Ofreq","percOfgenes")
  maxcluslen <- max(nchar(clusterdf$clusterSeq2))
  sizeDF <- data.frame(seq(1,maxcluslen,1),rep(0,maxcluslen))
  names(sizeDF) <- c("ClusterLen","Tfreq")
  sizeDF$Lfreq <- 0
  sizeDF$Ofreq <- 0
  sizeDF$percOfgenes <- 0
  for (i in 1: nrow(sizeDF)) {
    sizeDF$Tfreq[i] <- nrow(clusterdf[clusterdf$class=="T" & nchar(clusterdf$clusterSeq2) == i,])
    sizeDF$Lfreq[i] <- nrow(clusterdf[clusterdf$class=="L" & nchar(clusterdf$clusterSeq2) == i,])
    sizeDF$Ofreq[i] <- nrow(clusterdf[clusterdf$class=="O" & nchar(clusterdf$clusterSeq2) == i,])
    size_i_freq <- sizeDF$Tfreq[i] + sizeDF$Lfreq[i] + sizeDF$Ofreq[i]
    gene_num <- size_i_freq * i 
    sizeDF$percOfgenes[i] <- round((gene_num/nrow(genefile))*100,digits = 0)
  }
  sizeDF2 <- melt(sizeDF, id = c("ClusterLen","percOfgenes"))
  # Sort the data by ClusterLen and variable
  # Calculate the cumulative sum of the variable value for each cluster
  # Create the plot
  sizeDF2$ClusterLen <- factor(sizeDF2$ClusterLen)
  sizeDF2$OLTorder <- 0
  sizeDF2[sizeDF2$variable == "Ofreq", ]$OLTorder <- 0
  sizeDF2[sizeDF2$variable == "Lfreq", ]$OLTorder <- 1
  sizeDF2[sizeDF2$variable == "Tfreq", ]$OLTorder <- 2
  sizeDF2_sorted <-
    sizeDF2[order(sizeDF2$ClusterLen, sizeDF2$OLTorder), ]
  df_cumsum <- ddply(sizeDF2_sorted, "ClusterLen",
                     transform, label_ypos = cumsum(value))
  df_cumsum$label_ypos2 <- ""
  df_cumsum$label2 <- ""
  for (i in 1:nrow(df_cumsum)) {
    if (i %% 3 == 0)
    {
      df_cumsum$label_ypos2[i] <- toString(df_cumsum$label_ypos[i])
      df_cumsum$label2[i] <-
        paste(toString(df_cumsum$percOfgenes[i]), "%", sep = "")
    }
    else
    {
      df_cumsum$label_ypos2[i] = ""
      df_cumsum$label2[i] = ""
    }
  }
  df_cumsum$label1 <- ""
  df_cumsum$label1 <- df_cumsum$value
  df_cumsum[df_cumsum$value < 15,]$label1 <- ""
  ggplot(data = df_cumsum, aes(x = ClusterLen, y = value, fill = variable)) +
    geom_bar(stat = "identity") +
    geom_text(
      aes(y = label_ypos, label = label1),
      vjust = 1.3,
      color = "gray20",
      size = 4
    ) +
    scale_fill_brewer(palette = "Paired") + geom_text(aes(y = as.integer(label_ypos2), label =
                                                            label2),
                                                      vjust = -0.3,
                                                      size = 5,color = "darkgreen") + 
    scale_fill_manual(values = c("#00AFBB", "#E7B800", "lightgray"),name = "Genomes",labels = c("Trypanosoma", "Leishmania", "Others")) +
    theme_minimal() + xlab("Cluster Length") + ylab("Frequency")
  
#}
```
\newpage
\textbf{5. missing.genes.visualization\(\)}\
```{r ,echo=FALSE, warning=FALSE}
   genefile <- annotate.final.geneset.round1()
  func_classes <- c("A", "R", "N", "D" ,"C", "Q", "E", "G", "H" ,"I" ,"L", "K" ,"M" ,"X", "F", "P", "S" ,"T", "W" ,"Y", "V")
  genomes <- names(table(genefile$sourceOrg))
  missing_classdf <- data.frame(genomes, rep(0, length(genomes)))
  names(missing_classdf) <- c("genomes", "percent_21")
  missing_classdf$missing <- ''
  missing_classdf$genomes <- as.character(missing_classdf$genomes)
  missing_classdf$clade <- ""
  for (i in 1:nrow(missing_classdf)) {
    clusternames <- missing_classdf$genome[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      missing_classdf$clade[i] <- "Lenriettii"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      missing_classdf$clade[i] <- "AfricanT"
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      missing_classdf$clade[i] <- "AmericanT"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      missing_classdf$clade[i] <- "Leptospira"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      missing_classdf$clade[i] <- "Lmajor"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      missing_classdf$clade[i] <- "Linfantum"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      missing_classdf$clade[i] <- "LMexicana"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      missing_classdf$clade[i] <- "LViannia"
  }
  
  missing_classdf$Ycol <- "0"
  for (i in 1:nrow(missing_classdf)) {
    included_classes <-
      names(table(genefile[genefile$sourceOrg == missing_classdf$genomes[i], ]$genefunc))
    missingClasses <- setdiff(func_classes, included_classes)
    if (length(missingClasses) == 0)
      missing_classdf$percent_21[i] <- 21
    else
    {
      missing_classdf$missing[i] <- paste(missingClasses, collapse = " ")
      missing_classdf$percent_21[i] <- 21 - length(missingClasses)
    }
    if (missing_classdf$genomes[i] == "TcruziTulacl2")
    {
      missing_classdf$genomes[i] <- "*TcruziTulacl2"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziSylvioX10-1-2012")
    {
      missing_classdf$genomes[i] <- "*TcruziSylvioX10-1-2012"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziSylvioX10-1")
    {
      missing_classdf$genomes[i] <- "*TcruziSylvioX10-1"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziJRcl4")
    {
      missing_classdf$genomes[i] <- "*TcruziJRcl4"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziEsmeraldo")
    {
      missing_classdf$genomes[i] <- "*TcruziEsmeraldo"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruzicruziDm28c")
    {
      missing_classdf$genomes[i] <- "*TcruzicruziDm28c"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziCLBrenerNon-Esmeraldo-like")
    {
      missing_classdf$genomes[i] <- "*TcruziCLBrenerNon-Esmeraldo-like"
      missing_classdf$Ycol[i] <- "1"
    }
    if (missing_classdf$genomes[i] == "TcruziCLBrenerEsmeraldo-like")
    {
      missing_classdf$genomes[i] <- "*TcruziCLBrenerEsmeraldo-like"
      missing_classdf$Ycol[i] <- "1"
    }
  }
  missing_classdf[missing_classdf$genomes == "TcruziDm28c", ]$Ycol <-
    "1"
  missing_classdf[missing_classdf$genomes == "TcruziDm28c", ]$missing <-
    "   Y"
  missing_classdf$extralabel <- ""
  missing_classdf[missing_classdf$genomes == "TcruziDm28c", ]$extralabel <-
    "F"
  clades <- c("Lmajor","Linfantum","LMexicana","LViannia","Lenriettii","Leptospira","AmericanT","AfricanT")
  missing_classdf <- missing_classdf[order(match(missing_classdf$clade, clades)),]
  missing_classdf$color <- "0"
  missing_classdf[missing_classdf$clade=="Linfantum",]$color <- "1"
  missing_classdf[missing_classdf$clade=="LViannia",]$color <- "1"
  missing_classdf[missing_classdf$clade=="Leptospira",]$color <- "1"
  missing_classdf[missing_classdf$clade=="AfricanT",]$color <- "1"

  # add star to the names of 8 genomes
  # "#56B4E9"
  missing_classdf$genomes <-
    factor(missing_classdf$genomes, levels = missing_classdf$genomes)
  missing_classdf[missing_classdf$Ycol==1,]$Ycol <- "red"
  missing_classdf[missing_classdf$Ycol==0,]$Ycol <- "black"
  p <-
    ggplot(data = missing_classdf,
           aes(x = genomes, y = percent_21, fill = missing_classdf$color)) +
    geom_bar(stat = "identity")  + scale_fill_manual(values = c("gray", "white")) + theme(
      axis.text.x = element_text(
        angle = 90,
        hjust = 1,
        size = 12
       # text = element_text(size = 14)
      ),
      plot.title = element_text(hjust = 0.5,size = 15),
      plot.subtitle = element_text(hjust = 0.5),
      axis.title.y = element_text(size = 14),
      legend.position = "none"
    ) +
    labs(
      title =
        "Annotated tRNA Functional Classes of 46 TriTryp Genomes Including 3571 tRNA Genes for which TRNAscan-SE2 and Aragorn Functional Annotations Agree",
      subtitle = "10 intron-containing Tyr genes in 9 AmericanT genomes were annotated following tRNAscan-SE2\n",
      x =
        "Genome",
      y = "Number of Annotated tRNA Classes (total=21)"
    ) + geom_text(
      aes(label =
            missing, color = missing_classdf$Ycol),
      vjust = 0.5,
      angle = 90,
      hjust = -0.1,
      size = 4.5
    ) + geom_text(
      aes(label =
            extralabel),
      vjust = 0.5,
      angle = 90,
      hjust = -0.1,
      size = 4.5
    ) + scale_color_manual(
      name = "Ycol",
      labels = c("red", "black"),
      values = c("red" = "red", "black" = "black")
    )
  
  p + annotate("rect", xmin = 0.6, xmax = 8.3, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 0.6, xmax = 0.6, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 8.3, xmax = 8.3, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") + 
    
    annotate("rect", xmin = 8.5, xmax = 11.4, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 8.5, xmax = 8.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 11.4, xmax = 11.4, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") + 
    
    annotate("rect", xmin = 11.6, xmax = 13.5, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 11.6, xmax =11.6, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 13.5, xmax = 13.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    
    annotate("rect", xmin = 13.7, xmax = 17.6, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 13.7, xmax =13.7, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 17.6, xmax = 17.6, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    
    annotate("rect", xmin = 17.8, xmax = 19.6, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 17.8, xmax =17.8, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 19.6, xmax = 19.6, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    
    annotate("rect", xmin = 19.8, xmax = 22.5, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 19.8, xmax =19.8, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 22.5, xmax = 22.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    
    annotate("rect", xmin = 22.8, xmax = 33.5, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 22.8, xmax =22.8, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 33.5, xmax = 33.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    
    annotate("rect", xmin = 33.7, xmax = 39.5, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 33.7, xmax =33.7, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 39.5, xmax = 39.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
   
    annotate("rect", xmin = 39.7, xmax = 46.5, ymin = 21.5, ymax =21.5, alpha=1,colour = "black") + 
    annotate("rect", xmin = 39.7, xmax =39.7, ymin = 21.2, ymax =21.5, alpha=1,colour = "black") +
    annotate("rect", xmin = 46.5, xmax = 46.5, ymin = 21.2, ymax =21.5, alpha=1,colour = "black")+
    annotate(geom="text", label = "Lmajor", y=22.1, x=4.5)  + 
    annotate(geom="text", label = "Linfantum", y=22.1, x=10) +
    annotate(geom="text", label = "LMexicana", y=22.1, x=12.6) + 
    annotate(geom="text", label = "LViannia", y=22.1, x=16) +
    annotate(geom="text", label = "Lenriettii", y=22.1, x=18.7) + 
    annotate(geom="text", label = "Leptospira", y=22.1, x=21.1) + 
    annotate(geom="text", label = "AmericanT", y=22.1, x=28) +
    annotate(geom="text", label = "AfricanT", y=22.1, x=37) +
    annotate(geom="text", label = "Excluded", y=22.1, x=44)
    ggsave("/home/fatemeh/TriTrypGenome/MissingClasses.tiff", device = "tiff", width = 15,height = 8.5,
         dpi = 300)
```
\newpage
\textbf{6. gene.nuc.composition\(\)}\
```{r ,echo=FALSE, warning=FALSE}
prepare.tsfm.input <- function(Anotated_genefile){
  library(gsubfn)
  library(ggplot2)
  library(seqinr)
  #table(Anotated_genefile[Anotated_genefile$sourceOrg=="TcruziDm28c",]$genefunc)
  # table(Anotated_genefile$foundby)
  # ara both  tse 
  # 36 3579    1 
  Anotated_genefile <- annotate.final.geneset.round1()
  tsfmInput <- Anotated_genefile[Anotated_genefile$genefunc!="#" & Anotated_genefile$genefunc != "Z",]
  # table(tsfmInput$foundby)
  # ara both 
  # 36 3469
  # remove genes of genomes TrangeliSC58(6 genes) and TcruziCLBrener(11 genes)
  lowqualityGenomes <- tsfmInput$sourceOrg=="TrangeliSC58" | tsfmInput$sourceOrg=="TcruziCLBrener"
  tsfmInput <- tsfmInput[!lowqualityGenomes,]
  #ara both 
  #34 3454
  # 3488 genes left to be alined
  resultpath <- "/home/fatemeh/TriTrypGenome/GeneAnnotation/"
  write.table(tsfmInput,col.names = TRUE,file = paste(resultpath,"tsfm_input_geneset.txt",sep = ""))
  tsfmInput
}
  # in sup table we use first round gene annotaion (with only score filtering) 
  # but for table 1 we use the tsfm gene dataset which has only 40 genomes. 
  # genefile <- prepare.tsfm.input()
  #____________________________________________________
  # script to generate Sup table 5
  #____________________________________________________
  # table 5 is in dataframe comp_df and table 1 in clade_Gene_Comp_df
  genefile <- annotate.final.geneset.round1()
  comp_df <- data.frame(table(genefile$sourceOrg))
  names(comp_df)<- c("genome","genes")
  comp_df$Gperc <- 0
  comp_df$Cperc <- 0
  comp_df$Tperc <- 0
  for (i in 1:nrow(comp_df)) {
    genes <- paste(genefile[genefile$sourceOrg==comp_df$genome[i],]$genesequence,sep = "",collapse = "")
    comp_df$Gperc[i] <- Gpercentage(genes)
    comp_df$Cperc[i] <- Cpercentage(genes)
    comp_df$Tperc[i] <- Tpercentage(genes)
  }
  comp_df$clade <- ""
  for (i in 1:nrow(comp_df)) {
    clusternames <- comp_df$genome[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      comp_df$clade[i] <- "Lenriettii"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      comp_df$clade[i] <- "AfricanT"
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      comp_df$clade[i] <- "AmericanT"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      comp_df$clade[i] <- "Leptospira"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      comp_df$clade[i] <- "Lmajor"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      comp_df$clade[i] <- "Linfantum"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      comp_df$clade[i] <- "LMexicana"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      comp_df$clade[i] <- "LViannia"
  }
  comp_df <- comp_df[order(comp_df$clade),]
  comp_df$Gperc <- round(comp_df$Gperc,digits = 1)
  comp_df$Cperc <- round(comp_df$Cperc,digits = 1)
  comp_df$Tperc <- round(comp_df$Tperc,digits = 1)
  
  # ____________________________________________________
  # script to generate Table 1
  #____________________________________________________
  custom.boot <- function(times, data) {
    boot <- mean(replicate(times, sd(sample(data, replace=T))/sqrt(length(data))))
    boot
  }
  genefile <- prepare.tsfm.input()
  genefile$clade <- ""
  for (i in 1:nrow(genefile)) {
    clusternames <- genefile$sourceOrg[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      genefile$clade[i] <- "Lenriettii"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      genefile$clade[i] <- "AfricanT"
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      genefile$clade[i] <- "AmericanT"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      genefile$clade[i] <- "Leptospira"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      genefile$clade[i] <- "Lmajor"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      genefile$clade[i] <- "Linfantum"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      genefile$clade[i] <- "LMexicana"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      genefile$clade[i] <- "LViannia"
  }
  Genome_comp_df <- data.frame(table(genefile$sourceOrg))
  names(Genome_comp_df) <- c("genome","genes")
  Genome_comp_df$Gperc <- 0
  Genome_comp_df$Cperc <- 0
  Genome_comp_df$Tperc <- 0
  for (i in 1:nrow(Genome_comp_df)) {
    genes <- paste(genefile[genefile$sourceOrg==Genome_comp_df$genome[i],]$genesequence,sep = "",collapse = "")
    Genome_comp_df$Gperc[i] <- Gpercentage(genes)
    Genome_comp_df$Cperc[i] <- Cpercentage(genes)
    Genome_comp_df$Tperc[i] <- Tpercentage(genes)
  }
  Genome_comp_df$clade <- "non"
  for (i in 1:nrow(Genome_comp_df)) {
    clusternames <- Genome_comp_df$genome[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      Genome_comp_df$clade[i] <- "Lenriettii"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      Genome_comp_df$clade[i] <- "AfricanT"
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      Genome_comp_df$clade[i] <- "AmericanT"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      Genome_comp_df$clade[i] <- "Leptospira"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      Genome_comp_df$clade[i] <- "Lmajor"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      Genome_comp_df$clade[i] <- "Linfantum"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      Genome_comp_df$clade[i] <- "LMexicana"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      Genome_comp_df$clade[i] <- "LViannia"
  }
  Genome_comp_df <- Genome_comp_df[Genome_comp_df$clade!="non",]
  
  clade_comp_df <- data.frame(table(Genome_comp_df$clade))
  names(clade_comp_df)<- c("clade","genomes")
  clade_comp_df$TotalGp <- 0
  clade_comp_df$TotalCp <- 0
  clade_comp_df$TotalTp <- 0
  clade_comp_df$genesLen <- 0
  for (i in 1:nrow(clade_comp_df)) {
    genes <- paste(genefile[genefile$clade==clade_comp_df$clade[i],]$genesequence,sep = "",collapse = "")
    clade_comp_df$genesLen[i] <- nchar(genes)
    clade_comp_df$TotalGp[i] <- Gpercentage(genes)
    clade_comp_df$TotalCp[i] <- Cpercentage(genes)
    clade_comp_df$TotalTp[i] <- Tpercentage(genes)
    }
  clade_comp_df$AveGp <- 0
  clade_comp_df$AveCp <- 0
  clade_comp_df$AveTp <- 0
  clade_comp_df$genes <- 0
  clade_comp_df$SDgenes <- 0
  clade_comp_df$sdG <- 0
  clade_comp_df$sdC <- 0
  clade_comp_df$sdT <- 0
  
  for (i in 1:nrow(clade_comp_df)) {
    clade_comp_df$genes[i] <- sum(Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$genes)
    clade_comp_df$SDgenes[i] <- custom.boot(1000,Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$genes)
    clade_comp_df$AveGp[i] <- mean(Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Gperc)
    clade_comp_df$AveCp[i] <- mean(Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Cperc)
    clade_comp_df$AveTp[i] <- mean(Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Tperc)
    clade_comp_df$sdG[i] <- custom.boot(1000,Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Gperc)
    clade_comp_df$sdC[i] <- custom.boot(1000,Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Cperc)
    clade_comp_df$sdT[i] <- custom.boot(1000,Genome_comp_df[Genome_comp_df$clade==clade_comp_df$clade[i],]$Tperc)
  }
  
  clade_comp_df$TotalGp <- round(clade_comp_df$TotalGp,digits = 1)
  clade_comp_df$TotalCp <- round(clade_comp_df$TotalCp,digits = 1)
  clade_comp_df$TotalTp <- round(clade_comp_df$TotalTp,digits = 1)
  clade_comp_df$AveGp <- round(clade_comp_df$AveGp,digits = 1)
  clade_comp_df$AveCp <- round(clade_comp_df$AveCp,digits = 1)
  clade_comp_df$AveTp <- round(clade_comp_df$AveTp,digits = 1)
  clade_comp_df$sdG <- round(clade_comp_df$sdG,digits = 1)
  clade_comp_df$sdC <- round(clade_comp_df$sdC,digits = 1)
  clade_comp_df$sdT <- round(clade_comp_df$sdT,digits = 1)
  clade_comp_df$SDgenes <- round(clade_comp_df$SDgenes,digits = 1)
  clade_comp_df$AveGp <- paste(clade_comp_df$AveGp," (",clade_comp_df$sdG,")",sep = "")
  clade_comp_df$AveCp <- paste(clade_comp_df$AveCp," (",clade_comp_df$sdG,")",sep = "")
  clade_comp_df$AveTp <- paste(clade_comp_df$AveTp," (",clade_comp_df$sdG,")",sep = "")
  # names(clade_comp_df)[6] <- "AveGp (SD)"
  # names(clade_comp_df)[7] <- "AveCp (SD)"
  # names(clade_comp_df)[8] <- "AveTp (SD)"
  # excluded_genomes <- c("TrangeliSC58","TcruziCLBrener", "TtheileriEdinburgh", "EmonterogeiiLV88", "BayalaiB08-376", "PconfusumCUL13","LtarentolaeParrotTarII")
  # good <- !(comp_df$genome %in% excluded_genomes)
  # CIF_genomes <- comp_df[good,]
  # CIF_genomes <- CIF_genomes[order(CIF_genomes$genes),]
  comp_df
```
\newpage
\textbf{7. prepare.tsfm.input\(\)}\
\textbf{Input}: output of function annotate.final.geneset.round1()\
\textbf{Output}: \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/GeneAnnotation/tsfm_input_geneset.txt}{\textbf{tsfm\_input\_geneset.txt}}\
Read the annotated gene file, removes \textbf{genes with ambiguty} marked as # (35 genes), genes with function \textbf{Sec}s (76 genes) and genes from two genomes genes of genomes \textbf{TrangeliSC58} and \textbf{TcruziCLBrener} and writes the selected genes in file tsfm_input_geneset.txt (17 genes). tsfm_input_geneset.txt file should be used for further alignment for the purpose of creating CIFs.\
We have \textbf{3478} genes left in this file to be aligned\
```{r pressure, echo=FALSE, include=FALSE}
  tsfmInput <- Anotated_genefile[Anotated_genefile$genefunc!="#" & Anotated_genefile$genefunc != "Z",]
  # table(tsfmInput$foundby)
  # ara both 
  # 36 3459
  # remove genes of genomes TrangeliSC58(6 genes) and TcruziCLBrener(11 genes)
  lowqualityGenomes <- tsfmInput$sourceOrg=="TrangeliSC58" | tsfmInput$sourceOrg=="TcruziCLBrener"
  tsfmInput <- tsfmInput[!lowqualityGenomes,]
  # 3478 genes left to be aligned
  resultpath <- "/home/fatemeh/TriTrypGenome/GeneAnnotation/"
  write.table(tsfmInput,col.names = TRUE,file = paste(resultpath,"tsfm_input_geneset.txt",sep = ""))
```
\
\textbf{6. write.tsfm.input.in.fasta()}\
This function will write the tsfm\_input\_geneset.txt file into a fasta format: \href{https://github.com/fhadinezhadUC/TriTrypGenome/blob/master/final_CIF_geneset.fasta}{\textbf{final\_CIF\_geneset.fasta}}\
\
\
```{r ,echo=FALSE, warning=FALSE}
#write.tsfm.input.in.fasta <- function() {
  # this function will read the prepared gene file as input for from tsfm_input_geneset.txt
  # For genes found by ony ara, it will add _ara<genemodel> at the end of gene id (36 genes)
  # keep all the information needed to make a fasta file for doing the alignment
  
  filepath <- "/home/fatemeh/TriTrypGenome/GeneAnnotation/tsfm_input_geneset.txt"
  genefile <-
    read.table(filepath, header = TRUE, colClasses = "character")
  araonly <- genefile$foundby == "ara"
  genefile$geneid[araonly] <-
    paste(genefile$geneid[araonly], "_ara", genefile$genefunc[araonly], sep = "")
  genefile$geneid[!araonly] <-
    paste(genefile$geneid[!araonly], "_", genefile$genefunc[!araonly], sep = "")
  
  geneseq <- character(length = length(genefile$geneid))
  geness <- character(length = length(genefile$geneid))
  geneDF_fasta <-
    data.frame(
      genefile$geneid,
      genefile$tsegeneseq,
      genefile$arageneseq,
      genefile$tsegeness,
      genefile$arageness,
      genefile$foundby
    )
  names(geneDF_fasta) <-
    c("GeneID",
      "GeneSeqTse",
      "GeneSeqAra",
      "SSTSE",
      "SSARA",
      "foundby")
  
  genefile$arageneseq <- as.character(genefile$arageneseq)
  genefile$tsegeneseq <- as.character(genefile$tsegeneseq)
  geneDF_fasta$GeneID <- as.character(geneDF_fasta$GeneID)
  geneDF_fasta$GeneSeqTse <- as.character(geneDF_fasta$GeneSeqTse)
  geneDF_fasta$GeneSeqAra <- as.character(geneDF_fasta$GeneSeqAra)
  geneDF_fasta$SSARA <- as.character(geneDF_fasta$SSARA)
  geneDF_fasta$SSTSE <- as.character(geneDF_fasta$SSTSE)
  
  geneDF_fasta$GeneSeq <- ""
  geneDF_fasta$GeneSeq <- geneDF_fasta$GeneSeqTse
  geneDF_fasta[geneDF_fasta$foundby=="ara",]$GeneSeq <- toupper(geneDF_fasta[geneDF_fasta$foundby=="ara",]$GeneSeqAra)
  write.fwf(
    data.frame(
      paste(">", geneDF_fasta$GeneID, sep = ""),
      geneDF_fasta$GeneSeq
    ),
    file = "/home/fatemeh/TriTrypGenome/final_CIF_geneset.fasta",
    sep = "\n",
    colnames = FALSE
  )
  
#  geneDF_fasta
#}
```
\textbf{The gene set tsfm\_input\_geneset.txt is passed to the script TriTrypAlignment.R to be aligned with Human tRNA genes.}\
\
\
\textbf{TriTrypAlignment.R}\
Alignment Steps:\
\begin{itemize}
\item[1.] Genes from \textbf{tsfm\_input\_geneset.txt} are read and functional classes are added at the end of the geneID\
\item[2.] Variable arms are removed based on reported secondary structure from genefinders\
\item[3.] Gene introns are removed based in the secondary sctructure\
\item[4.] the result is merged with the Human tRNA genes (the headers for Homo genes are also updated) and the result is saved in file coveainput.fasta.\
\item[5.] coveainput.fasta is aligned to the Eukaryote model using covea\
\item[6.] the result (Aligned\_TriTryp\_Homo.covea) is edited based on the fallowing criteria in order:\
\begin{itemize}
   \item[a.]  sites that have more than 98% gap are removed\
   \item[b.] sequences with more than 3 gaps are removed\
   \item[c.] sequences with two or more gaps next to eachother are removed\
\end{itemize}
\item[7.] the alignment result is saved as fasta file in Aligned\_TriTryp\_Homo.fasta, with secondary structure saved as Aligned\_TriTryp\_Homo\_structfile.txt\
\end{itemize}

\
\
\textbf{SplitAlignedGenes.R}
\
The alignment Result Aligned_TriTryp_Homo.fasta will be passed to this script to be splitted either by genome, or clusters of genomes.\
These genomes were not included in any of the clusters: TrangeliSC58, TcruziCLBrener, TtheileriEdinburgh, EmonterogeiiLV88, BayalaiB08-376, PconfusumCUL13\
The result fasta file for each genome is saved as a file in tsfm/input folder.\
The missing functional class for each genome or cluster of genomes is visualized as a bar plot.\
\
\

```{r , echo=FALSE, include=FALSE}
split_tRNAgene_with_clustername <- function() {
  resultpath <- "/home/fatemeh/TriTrypGenome/tsfm/sites72/"
  fastafile <-
    read.table(
      "/home/fatemeh/TriTrypGenome/GeneAnnotation/Aligned_TriTryp_Homo.fasta",
      sep = "\n"
    )
  fastafile <- as.character(fastafile$V1)
  sequences <-
    as.character(fastafile[seq(2, length(fastafile), 2)])
  headers <- as.character(fastafile[seq(1, length(fastafile), 2)])
  headers <- gsub("_ara", "_", headers)
  headers <- gsub("\\s", "", headers)
  
  headers2 <-
    lapply(X = headers , function(X)
      gsub("Homo", "HOMO_", X, fixed = TRUE))
  headers2 <- as.character(headers2)
  
  genomeName <-
    lapply(X = headers2 , function(X)
      gsub(">", "", X, fixed = TRUE))
  genomeName <- as.character(genomeName)
  
  genomeName2 <-
    lapply(X = genomeName, function(X)
      unlist(strsplit(X, split = "_"))[1])
  genomeName2 <- as.character(genomeName2)
  
  # we will split genomes based on clusters of genomes
  # add another column "clustername" to tRNAdf to assign the new names to each cluster of genomes:
  clusNames <- character(length = length(headers))
  clusNames <- genomeName2
  for (i in 1:length(clusNames)) {
    clusternames <- clusNames[i]
    if (clusternames == "LspMARLEM2494" |
        clusternames == "LenriettiiLEM3045")
      clusNames[i] <- "LenriettiComplex"
    if (clusternames == "TbruceigambienseDAL972" |
        clusternames == "TbruceiLister427" |
        clusternames == "TbruceiTREU927" |
        clusternames == "TevansiSTIB805" |
        clusternames == "TcongolenseIL3000" |
        clusternames == "TvivaxY486")
      clusNames[i] <- "AfricanTrypanosome"
    #clusternames == "TcruziCLBrener" clusternames == "TrangeliSC58" removed
    if (clusternames == "TgrayiANR4" |
        clusternames == "TcruziCLBrenerEsmeraldo-like" |
        clusternames == "TcruziCLBrenerNon-Esmeraldo-like" |
        clusternames == "TcruzicruziDm28c" |
        clusternames == "TcruziDm28c" |
        clusternames == "TcruziEsmeraldo" |
        clusternames == "TcruziJRcl4" |
        clusternames == "TcruzimarinkelleiB7" |
        clusternames == "TcruziSylvioX10-1" |
        clusternames == "TcruziSylvioX10-1-2012" |
        clusternames == "TcruziTulacl2")
      # clusternames == "TtheileriEdinburgh"
      clusNames[i] <- "AmericanTrypanosome"
    if (clusternames == "CfasciculataCfCl" |
        clusternames == "LseymouriATCC30220" |
        clusternames == "LpyrrhocorisH10")
      clusNames[i] <- "Leishmania1"
    if (clusternames == "LmajorFriedlin" |
        clusternames == "LmajorLV39c5" |
        clusternames == "LmajorSD75" |
        clusternames == "LturanicaLEM423" |
        clusternames == "LarabicaLEM1108" |
        clusternames == "LtropicaL590" |
        clusternames == "LaethiopicaL147" |
        clusternames == "LgerbilliLEM452")
      clusNames[i] <- "Leishmania3"
    if (clusternames == "LdonovaniBHU1220" |
        clusternames == "LdonovaniBPK282A1" |
        clusternames == "LinfantumJPCM5")
      clusNames[i] <- "Leishmania4"
    if (clusternames == "LamazonensisMHOMBR71973M2269" |
        clusternames == "LmexicanaMHOMGT2001U1103")
      clusNames[i] <- "Leishmania2"
    if (clusternames == "LbraziliensisMHOMBR75M2904" |
        clusternames == "LbraziliensisMHOMBR75M2903" |
        clusternames == "LpanamensisMHOMPA94PSC1" |
        clusternames == "LpanamensisMHOMCOL81L13")
      clusNames[i] <- "Leishmania5"
  }
  
  functionalclasses <-
    lapply(X = headers, function(X)
      unlist(strsplit(X, split = "_"))[length(unlist(strsplit(X, split = "_")))])
  
  functionalclasses <- as.character(functionalclasses)
  
  tRNAdf <-
    data.frame(genomeName2,
               sequences,
               headers2,
               functionalclasses,
               clusNames)
  names(tRNAdf) <-
    c("genomename", "sequence", "headers", "funclass", "clusNames")
  
  genome_list = split(tRNAdf, f = tRNAdf$clusNames)
  
  # make a file for each genome and put the genomes in there
  for (i in 1:length(genome_list)) {
    # write the DF in a file to be processed by the bash
    filepath <-
      paste(resultpath,
            names(genome_list[i]),
            ".fasta",
            sep = "")
    
    write.fwf(
      data.frame(genome_list[[i]]$headers,
                 genome_list[[i]]$sequence),
      filepath,
      sep = "\n",
      colnames = FALSE
    )
  }
 # vidualization(genome_list, )
}
```

```{r , echo=FALSE, include=FALSE}
#s  resultpath <- "/home/fatemeh/TriTrypGenome_project/tsfm/input1/"
  fastafile <-
    read.table(
      "/home/fatemeh/TriTrypGenome_project/GeneAnnotation/Aligned_TriTryp_Homo.fasta",
      sep = "\n"
    )
  fastafile <- as.character(fastafile$V1)
  sequences <-
    as.character(fastafile[seq(2, length(fastafile), 2)])
  headers <- as.character(fastafile[seq(1, length(fastafile), 2)])
  headers <- gsub("_ara", "_", headers)
  headers <- gsub("\\s", "", headers)
  
  headers2 <-
    lapply(X = headers , function(X)
      gsub("Homo", "HOMO_", X, fixed = TRUE))
  headers2 <- as.character(headers2)
  
  genomeName <-
    lapply(X = headers2 , function(X)
      gsub(">", "", X, fixed = TRUE))
  genomeName <- as.character(genomeName)
  
  genomeName2 <-
    lapply(X = genomeName, function(X)
      unlist(strsplit(X, split = "_"))[1])
  genomeName2 <- as.character(genomeName2)
  
  
  functionalclasses <-
    lapply(X = headers, function(X)
      unlist(strsplit(X, split = "_"))[length(unlist(strsplit(X, split = "_")))])
  
  functionalclasses <- as.character(functionalclasses)
  
  tRNAdf <-
    data.frame(genomeName2, sequences, headers2, functionalclasses)
  names(tRNAdf) <-
    c("genomename", "sequence", "headers", "funclass")
  
  genome_list = split(tRNAdf, f = tRNAdf$genomename)
  
  # make a file for each genome and put the genomes in there
  for (i in 1:length(genome_list)) {
    # write the DF in a file to be processed by the bash
    filepath <-
      paste(resultpath,
            names(genome_list[i]),
            ".fasta",
            sep = "")
    
    write.fwf(
      data.frame(genome_list[[i]]$headers,
                 genome_list[[i]]$sequence),
      filepath,
      sep = "\n",
      colnames = FALSE
    )
    
  }
#  vidualization(genome_list, resultpath)
#}
```

```{r , echo=FALSE, include=FALSE}
  tRNAcountdf <- data.frame(names(genome_list))
  tRNAcountdf$counts <- 0
  names(tRNAcountdf) <- c("genome", "tRNAcounts")
  for (i in 1:length(genome_list)) {
    tRNAcountdf[tolower(tRNAcountdf$genome) == tolower(names(genome_list[i])),]$tRNAcounts <-
      nrow(genome_list[[i]])
  }
  
  tRNAcountdf <- tRNAcountdf[order(tRNAcountdf$tRNAcounts), ]
  tRNAcountdf$genome <-
    factor(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  tRNAcountdf$genome <-
    as.character(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  
  func_classes <-
    c(
      "A",
      "R",
      "N" ,
      "D",
      "C" ,
      "Q",
      "E",
      "G",
      "H",
      "I",
      "L",
      "K",
      "M" ,
      "X",
      "F",
      "P",
      "S",
      "T",
      "W",
      "Y" ,
      "V"
    )
  tRNAcountdf$percent_21 <- 0
  tRNAcountdf$missing <- ''
  for (i in 1:length(genome_list)) {
    if (tolower(names(genome_list[i])) != "homo")
    {
      countsdf <-
        as.data.frame(table(as.character(genome_list[[i]]$funclass)))
      tRNAcountdf[tRNAcountdf$genome == names(genome_list[i]), ]$missing <-
        paste(setdiff(func_classes, as.character(countsdf$Var1)), collapse = " ")
      diffset <-
        setdiff(func_classes, as.character(countsdf$Var1))
      fcount <- 21 - length(diffset)
      tRNAcountdf[tRNAcountdf$genome == names(genome_list[i]), ]$percent_21 <-
        (fcount / 21) * 100
    }
    else
      tRNAcountdf[tolower(tRNAcountdf$genome) == tolower(names(genome_list[i])), ]$percent_21 <-
        100
    
  }
  tRNAcountdf$genome <-
    factor(tRNAcountdf$genome, levels = tRNAcountdf$genome)
  p <- ggplot(data = tRNAcountdf, aes(x = genome, y = percent_21)) +
    geom_bar(stat = "identity", fill = "#56B4E9")  + theme(
      axis.text.x = element_text(angle = 90, hjust = 1),
      plot.title = element_text(hjust = 0.5) ,
      text = element_text(size = 8)
    ) +
    geom_text(
      aes(y = tRNAcountdf$percent_21, label = tRNAcountdf$tRNAcounts),
      vjust = 1.3,
      color = "gray20",
      size = 2
    ) +
    labs(title =
           "Percentage of 22 tRNA functional classes covered by each cluster",
         x =
           "Genome", y = "Percentage of tRNA functional classes(total=22)") + geom_text(
             aes(label =
                   missing),
             vjust = 0.5,
             angle = 90,
             hjust = -0.1,
             color = "blueviolet",
             size = 2
           )
  p
  # ggsave(
  #   paste(plotpath, "funcPerc_clustered2.png", sep = ""),
  #   width = 14,
  #   height = 9
  # )
```
\
\
\
Fasta gene files will be splitted based on tRNA functional class by running script splitFuncClass.sh.

